### PR TITLE
fix: enhance JSON handling and error reporting in profiling functions

### LINKED ISSUE #387: P0: make large-file distribution statistics deterministic and honest across engines
Parent: #371.

## Severity

P0 product-contract bug: identical input can produce different numeric output across runs.

## Dogfooding reproduction

A 500,000-row numeric CSV was profiled repeatedly:

| engine/run | `amount.mean` |
|---|---:|
| incremental | 383.1112 |
| auto | 386.1255 |
| columnar | 384.5769 |

The true full-column mean is the columnar value. The incremental path tracks exact Welford min/max/mean/variance online, but `profile_from_stats_with_hints()` discards those aggregates and recomputes the public `NumericStats` from a random 10,000-value reservoir. The reservoir was seeded from OS randomness, so identical runs selected different values.

The immediate dogfooding fix gives the internal reservoir a fixed seed, restoring repeatability. The larger contract remains: exact online aggregates should not be replaced with sampled ones, and sampled distribution metrics must be clearly marked.

## Required behavior

- exact online fields (count, nulls, min, max, mean, variance/stddev where numerically valid) come from global accumulators
- sampled fields (median, quartiles, mode, skew/kurtosis, outlier estimates) are deterministic and marked approximate when the reservoir truncates
- engine differences are either eliminated or explicitly represented as confidence/provenance, never silent

## Acceptance criteria

- [ ] Profiling the same ordered source repeatedly yields byte-stable numeric report values.
- [ ] Streaming mean/min/max and variance use global online aggregates.
- [ ] Sample-derived metrics set `is_approximate=true` once input exceeds the retained sample.
- [ ] Sampling seed/policy is documented and stable or user-configurable.
- [ ] Cross-engine tests cover >10,000 rows and assert exact fields plus approximation metadata.
- [ ] Quality scores do not vary across identical repeated runs.
- [ ] Memory remains bounded.

## Likely files

- `crates/dataprof-runtime/src/streaming_stats.rs`
- `crates/dataprof-runtime/src/profile_builder.rs`
- `crates/dataprof-metrics/src/stats/numeric.rs`
- large parity regression tests

## Related

- #383 distinct-count approximation provenance
- #361 streaming guardrail epic

### PR DESCRIPTION
**Pull request by AndreaBozzo**
- Improved JSON serialization for nested structures in `_cell_to_str`.
- Added validation for duplicate keys in dictionary inputs to `_columns_from_dict`.
- Enhanced `_columns_from_records` to support max row limits and key sorting.
- Updated tests to cover new functionality and edge cases for JSON and dictionary inputs.

# Pull Request

**👋 Thank you for contributing to dataprof!** 

Please fill out this template to help us understand your changes.

## 📝 Summary

Brief description of the changes in this PR. What problem does it solve or what feature does it add?

## 🔧 Type of Change

Select the type(s) of change:

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 🚨 Breaking change (would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] ⚡ Performance improvement
- [ ] 🎨 Code refactoring (no functional changes)

## 📋 Changes Made

Describe the specific changes:

- Change 1
- Change 2
- Change 3

**Related Issue:** Closes #(issue number)

### How did you test this?

Describe your testing approach:

- [ ] Tested locally with `cargo test`
- [ ] Added new tests for this feature
- [ ] All existing tests pass
- [ ] Tested with sample data files
- [ ] Tested edge cases (empty files, large files, etc.)

### Test Commands Run

```bash
cargo test --all
cargo clippy --all --all-targets
cargo fmt --all -- --check
cargo build --release
```

Include any specific test scenarios or data used.

## 📖 Documentation

- [ ] Updated README.md if user-facing changes
- [ ] Added/updated inline code comments
- [ ] Updated docs/ folder if architectural changes
- [ ] No documentation changes needed

## ⚡ Performance Impact

- [ ] No performance impact
- [ ] Performance improved
- [ ] Performance may be affected (explain below)

**Details (if applicable):**

## 🚨 Breaking Changes

If this is a breaking change, describe:

- What breaks
- Migration path for users

**If no breaking changes, just check:** [ ] No breaking changes

## 📌 Additional Notes

Any additional context for reviewers:

- Screenshots/examples (if applicable)
- Related PRs or discussions
- Implementation notes

---

## ✅ Checklist

- [ ] Code builds without warnings
- [ ] No linting errors (`cargo clippy`)
- [ ] Code is formatted (`cargo fmt`)
- [ ] All tests pass
- [ ] Self-reviewed my code
- [ ] Comments added for complex logic
- [ ] No hardcoded secrets or sensitive data

**Need help?** See [CONTRIBUTING.md](../../CONTRIBUTING.md) for guidelines and [docs/](../../docs/) for architecture details.

### CODE REVIEW

File: python/dataprof/__init__.py [hunk 1]
```diff
@@ -222,19 +222,29 @@ def _cell_to_str(value: object) -> str | None:
         return None
     if isinstance(value, str):
         return value
+    if isinstance(value, (dict, list)):
+        return json.dumps(value, ensure_ascii=False, separators=(",", ":"), sort_keys=True)
     return str(value)
```
[the remark this answers is not in the document]
Comment (AndreaBozzo) [reply python/dataprof/__init__.py:232]: Fixed in 6a70762. `_cell_to_str` now attempts compact JSON and falls back to `str(value)` on `TypeError`/`ValueError`, so a dict/list cell holding a datetime, set, or circular ref degrades gracefully instead of aborting profiling. Added a regression test (`test_container_cell_with_unserialisable_contents_falls_back_to_str`).

File: python/dataprof/__init__.py [hunk 2]
```diff
@@ -243,13 +253,41 @@ def _columns_from_dict(source: dict[Any, Any]) -> list[_Column]:
...
+        )
+        raise ValueError(
+            "list-of-dicts input: column keys collide after string conversion: "
+            f"{collisions!r}. Use distinct string column names."
+        )
+
+    return [
+        (name, [_cell_to_str(row.get(key)) for row in rows])
+        for key, name in zip(keys, normalized_names, strict=True)
+    ]
```
[the remark this answers is not in the document]
Comment (AndreaBozzo) [reply python/dataprof/__init__.py]: Fixed in 6a70762. `_columns_from_records` now uses two windows: key discovery stays at the first `max_rows` rows, while cells are materialized for `max_rows + 1` rows. The extra row is ignored by the Rust profiler (it analyzes only `max_rows`) but its presence is what lets Rust detect truncation via `num_rows < source_rows` — passing exactly `max_rows` rows would flip `source_exhausted` to True. This avoids stringifying dropped rows while preserving the truncation signal. The empty-records error is now based on the analyzed subset, so `profile([...], max_rows=0)` returns 0 rows instead of raising the misleading 'rows but no columns' error. Added `test_list_of_dicts_max_rows_zero_yields_no_rows_not_error`.
